src/Symfony/Component/HttpClient/Rfc9111HttpClient.php

```diff
@@ -0,0 +1,616 @@
+use Symfony\Component\HttpClient\Response\MockResponse;
+use Symfony\Component\HttpClient\Response\ResponseStream;
+use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+use Symfony\Contracts\Service\ResetInterface;
+
+/**
+ * Adds caching on top of an HTTP client (per RFC 9111).
+ *
+ * Known omissions / partially supported features per RFC 9111:
+ *   1. Range requests:
+ *     - All range requests ("partial content") are passed through and never cached.
+ *   2. stale-while-revalidate:
+ *     - There's no actual "background revalidation" for stale responses, they will
+ *       always be revalidated.
+ *   3. min-fresh, max-stale, only-if-cached:
+ *     - Request directives are not parsed; the client ignores them.
+ *
+ * @see https://www.rfc-editor.org/rfc/rfc9111
+ */
+class Rfc9111HttpClient implements HttpClientInterface, ResetInterface
+{
+    use AsyncDecoratorTrait {
+        stream as asyncStream;
+        AsyncDecoratorTrait::withOptions insteadof HttpClientTrait;
+    }
+    use HttpClientTrait;
+
+    private array $defaultOptions = self::OPTIONS_DEFAULTS;
+
+    /**
+     * @param HttpClientInterface      $client                           The decorated client
+     * @param TagAwareAdapterInterface $cache                            The cache to store responses
+     * @param array                    $defaultOptions                   The default options for the client
+     * @param bool                     $sharedCache                      Whether to share the cache between all instances of this client
+     * @param int|null                 $ttl                              The default TTL (in seconds) for the cache
+     * @param array<int>               $cachableStatusCodes              The status codes that are always cacheable
+     * @param array<int>               $conditionallyCachableStatusCodes The status codes that are cacheable if the response has a Cache-Control header or ETag
+     * @param array<string>            $cacheableMethods                 The HTTP methods that are cacheable
+     * @param array<string>            $unsafeMethods                    The HTTP methods that will trigger a cache invalidation
+     */
+    public function __construct(
+        HttpClientInterface $client,
+        private readonly TagAwareAdapterInterface $cache,
+        array $defaultOptions = [],
+        private readonly bool $sharedCache = true,
+        private readonly ?int $ttl = null,
+        private readonly array $cachableStatusCodes = [200, 203, 204, 206, 300, 301, 404, 405, 410, 414, 501],
+        private readonly array $conditionallyCachableStatusCodes = [302, 303, 307, 308],
+        private readonly array $cacheableMethods = ['GET', 'HEAD'],
+        private readonly array $unsafeMethods = ['POST', 'PUT', 'DELETE', 'PATCH'],
```

Comment:
why do we need to make those unsafe methods configurable ? The spec is the one defining which methods are unsafe. I don't see the value of increasing the complexity of the configuration by making them configurable (and making it possible to have a Rfc9111HttpClient that behaves in a way incompatible with the RFC 9111)

Reply:
https://github.com/symfony/symfony/pull/59576#discussion_r1925262705

> We could put as a class const, but I've came across APIs that proposed POST endpoints to let their users express complex search queries that wouldn't fit in a regular GET endpoint with url parameters, so I wanted to let the users be able to configure that.

Reply:
The default values are inline with what RFC 9111 defines, but I wanted to let the users be able to tweak them to match their odd cases (because if that's not the case, you can be sure they'll open issues for this here).

Reply:
the issue with that is that configurable `<cacheable-method>PUT</cacheable-method>` does not only make `PUT` cacheable. It also makes all RFC-cacheable methods uncacheable because the default value of a prototyped array node is not merged with the provided configuration in the Config component (and the default value of a PHP array argument is not merged either with the provided value).

I would suggest that the first version of this client does not have such configuration. Our existing CachingHttpClient does not expose such configuration either and I'm not aware of any request to add support for it. If we start without this feature, we can introduce it later. However, if we make it configurable now, it is very hard to _remove_ the feature later.

Reply:
I've removed all configuration related to status codes and http methods.